File: gradle-plugin/src/main/java/io/spine/embedcode/gradle/EmbedCodeTask.java
/*
 * Copyright 2026, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.embedcode.gradle;

import org.gradle.api.DefaultTask;
import org.gradle.api.GradleException;
import org.gradle.api.file.ConfigurableFileCollection;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.MapProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputDirectory;
import org.gradle.api.tasks.InputFile;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.Optional;
import org.gradle.api.tasks.PathSensitive;
import org.gradle.api.tasks.PathSensitivity;
import org.gradle.api.tasks.TaskAction;
import org.gradle.process.ExecOperations;
import org.gradle.work.DisableCachingByDefault;

import javax.inject.Inject;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

/** Runs Embed Code in either check or embed mode. */
@DisableCachingByDefault(because = "Embed Code checks or updates documentation files in place")
public abstract class EmbedCodeTask extends DefaultTask {

    /** Returns process execution without project access at execution time. */
    @Inject
    protected abstract ExecOperations getExecOperations();

    /** Returns the execution mode assigned by the plugin. */
    @Input
    public abstract Property<String> getMode();

    /** Returns the source root passed to {@code -code-path}. */
    @InputDirectory
    @Optional
    @PathSensitive(PathSensitivity.RELATIVE)
    public abstract DirectoryProperty getCodePath();

    /** Returns named source roots included in an internally generated configuration. */
    @Input
    public abstract MapProperty<String, String> getNamedSources();

    /** Returns named source directories with their producing task dependencies. */
    @InputFiles
    @PathSensitive(PathSensitivity.RELATIVE)
    public abstract ConfigurableFileCollection getNamedSourceDirectories();

    /** Returns the documentation root passed to {@code -docs-path}. */
    @InputDirectory
    @PathSensitive(PathSensitivity.RELATIVE)
    public abstract DirectoryProperty getDocsPath();

    /** Returns documentation include patterns passed to {@code -doc-includes}. */
    @Input
    public abstract ListProperty<String> getDocIncludes();

    /** Returns documentation exclude patterns passed to {@code -doc-excludes}. */
    @Input
    public abstract ListProperty<String> getDocExcludes();

    /** Returns the fragment separator passed to {@code -separator}. */
    @Input
    public abstract Property<String> getSeparator();

    /** Returns whether informational logging is enabled. */
    @Input
    public abstract Property<Boolean> getInfo();

    /** Returns whether panic stack traces are enabled. */
    @Input
    public abstract Property<Boolean> getStacktrace();

    /** Returns the installed platform executable. */
    @InputFile
    @PathSensitive(PathSensitivity.NONE)
    public abstract RegularFileProperty getExecutableFile();

    /** Returns the process working directory. */
    @Internal
    public abstract DirectoryProperty getWorkingDirectory();

    /** Executes Embed Code with arguments derived from the Gradle extension. */
    @TaskAction
    public void runEmbedCode() {
        Map<String, String> namedSources = new TreeMap<>(getNamedSources().get());
        boolean hasDirectSource = getCodePath().isPresent();
        boolean hasNamedSources = !namedSources.isEmpty();
        if (hasDirectSource == hasNamedSources) {
            throw new GradleException(
                    "Configure exactly one of `codePath` or `namedSource(...)` for Embed Code."
            );
        }

        List<String> arguments = new ArrayList<>();
        arguments.add("-mode=" + getMode().get());
        if (hasNamedSources) {
            arguments.add("-config-path=" + writeNamedSourceConfiguration(namedSources));
        } else {
            arguments.add("-code-path=" + getCodePath().get().getAsFile().getAbsolutePath());
            arguments.add("-docs-path=" + getDocsPath().get().getAsFile().getAbsolutePath());
            if (!getDocIncludes().get().isEmpty()) {
                arguments.add("-doc-includes=" + String.join(",", getDocIncludes().get()));
            }
            if (!getDocExcludes().get().isEmpty()) {
                arguments.add("-doc-excludes=" + String.join(",", getDocExcludes().get()));
            }
            arguments.add("-separator=" + getSeparator().get());
            arguments.add("-info=" + getInfo().get());
            arguments.add("-stacktrace=" + getStacktrace().get());
        }

        getExecOperations().exec(spec -> {
            spec.executable(getExecutableFile().get().getAsFile());
            spec.args(arguments);
            spec.setWorkingDir(getWorkingDirectory().get().getAsFile());
        });
    }

    /** Writes the generated configuration used when named source roots are configured. */
    private Path writeNamedSourceConfiguration(Map<String, String> namedSources) {
        Map<String, String> normalizedSources = new TreeMap<>();
        for (Map.Entry<String, String> source : namedSources.entrySet()) {
            Path path = Paths.get(source.getValue());
            if (!path.isAbsolute()) {
                path = getWorkingDirectory().get().getAsFile().toPath().resolve(path);
            }
            path = path.normalize().toAbsolutePath();
            if (!Files.isDirectory(path)) {
                throw new GradleException(
                        "Embed Code source `" + source.getKey() + "` is not a directory: " + path
                );
            }
            normalizedSources.put(source.getKey(), path.toString());
        }

        String json = createConfigurationJson(
                normalizedSources,
                getDocsPath().get().getAsFile().getAbsolutePath(),
                getDocIncludes().get(),
                getDocExcludes().get(),
                getSeparator().get(),
                getInfo().get(),
                getStacktrace().get()
        );
        Path configuration = getTemporaryDir().toPath().resolve("embed-code.json");
        try {
            Files.write(configuration, json.getBytes(StandardCharsets.UTF_8));
        } catch (IOException exception) {
            throw new GradleException(
                    "Could not write the generated Embed Code configuration to "
                            + configuration + '.',
                    exception
            );
        }
        return configuration;
    }

    /** Creates a JSON document accepted by Embed Code's YAML configuration parser. */
    static String createConfigurationJson(
            Map<String, String> namedSources,
            String docsPath,
            List<String> docIncludes,
            List<String> docExcludes,
            String separator,
            boolean info,
            boolean stacktrace
    ) {
        StringBuilder json = new StringBuilder();
        json.append("{\n  \"code-path\": [\n");
        int index = 0;
        for (Map.Entry<String, String> source : namedSources.entrySet()) {
            if (index > 0) {
                json.append(",\n");
            }
            json.append("    {\"name\": ");
            appendJsonString(json, source.getKey());
            json.append(", \"path\": ");
            appendJsonString(json, source.getValue());
            json.append('}');
            index++;
        }
        json.append("\n  ],\n  \"docs-path\": ");
        appendJsonString(json, docsPath);
        json.append(",\n  \"doc-includes\": ");
        appendJsonArray(json, docIncludes);
        json.append(",\n  \"doc-excludes\": ");
        appendJsonArray(json, docExcludes);
        json.append(",\n  \"separator\": ");
        appendJsonString(json, separator);
        json.append(",\n  \"info\": ").append(info);
        json.append(",\n  \"stacktrace\": ").append(stacktrace);
        json.append("\n}\n");
        return json.toString();
    }

    /** Appends a JSON array containing {@code values}. */
    private static void appendJsonArray(StringBuilder json, List<String> values) {
        json.append('[');
        for (int i = 0; i < values.size(); i++) {
            if (i > 0) {
                json.append(", ");
            }
            appendJsonString(json, values.get(i));
        }
        json.append(']');
    }

    /** Appends {@code value} as an escaped JSON string. */
    private static void appendJsonString(StringBuilder json, String value) {
        json.append('"');
        for (int i = 0; i < value.length(); i++) {
            char character = value.charAt(i);
            switch (character) {
                case '"':
                    json.append("\\\"");
                    break;
                case '\\':
                    json.append("\\\\");
                    break;
                case '\b':
                    json.append("\\b");
                    break;
                case '\f':
                    json.append("\\f");
                    break;
                case '\n':
                    json.append("\\n");
                    break;
                case '\r':
                    json.append("\\r");
                    break;
                case '\t':
                    json.append("\\t");
                    break;
                default:
                    if (character < 0x20) {
                        json.append(String.format(Locale.ROOT, "\\u%04x", (int) character));
                    } else {
                        json.append(character);
                    }
            }
        }
        json.append('"');
    }
}

File: gradle-plugin/src/main/kotlin/io/spine/embedcode/gradle/EmbedCodeTask.kt
/*
 * Copyright 2026, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.embedcode.gradle

import org.gradle.api.DefaultTask
import org.gradle.api.GradleException
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputDirectory
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity
import org.gradle.api.tasks.TaskAction
import org.gradle.process.ExecOperations
import org.gradle.work.DisableCachingByDefault
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.TreeMap
import javax.inject.Inject

/**
 * Runs Embed Code in either check or embed mode.
 */
@DisableCachingByDefault(because = "Embed Code checks or updates documentation files in place")
public abstract class EmbedCodeTask : DefaultTask() {

    /** Process execution without project access at execution time. */
    @get:Inject
    protected abstract val execOperations: ExecOperations

    /** The execution mode assigned by the plugin. */
    @get:Input
    public abstract val mode: Property<String>

    /** The source root passed to `-code-path`. */
    @get:InputDirectory
    @get:Optional
    @get:PathSensitive(PathSensitivity.RELATIVE)
    public abstract val codePath: DirectoryProperty

    /**
     * Named source roots included in an internally generated configuration.
     *
     * The absolute path values make this input machine-specific. The task intentionally
     * declares no outputs and disables caching because it checks or modifies documentation.
     */
    @get:Input
    public abstract val namedSources: MapProperty<String, String>

    /** Named source directories with their producing task dependencies. */
    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    public abstract val namedSourceDirectories: ConfigurableFileCollection

    /** The documentation root passed to `-docs-path`. */
    @get:InputDirectory
    @get:PathSensitive(PathSensitivity.RELATIVE)
    public abstract val docsPath: DirectoryProperty

    /** Documentation include patterns passed to `-doc-includes`. */
    @get:Input
    public abstract val docIncludes: ListProperty<String>

    /** Documentation exclude patterns passed to `-doc-excludes`. */
    @get:Input
    public abstract val docExcludes: ListProperty<String>

    /** The fragment separator passed to `-separator`. */
    @get:Input
    public abstract val separator: Property<String>

    /** Whether informational logging is enabled. */
    @get:Input
    public abstract val info: Property<Boolean>

    /** Whether panic stack traces are enabled. */
    @get:Input
    public abstract val stacktrace: Property<Boolean>

    /** The installed platform executable. */
    @get:InputFile
    @get:PathSensitive(PathSensitivity.NONE)
    public abstract val executableFile: RegularFileProperty

    /** The process working directory. */
    @get:Internal
    public abstract val workingDirectory: DirectoryProperty

    /**
     * Executes Embed Code with arguments derived from the Gradle extension.
     */
    @TaskAction
    public fun runEmbedCode() {
        val executionMode = mode.get()
        val processDirectory = workingDirectory.get().asFile
        logger.info(
            "Preparing Embed Code `{}` mode in `{}`.",
            executionMode,
            processDirectory,
        )
        val configuredSources = TreeMap(namedSources.get())
        val hasDirectSource = codePath.isPresent
        val hasNamedSources = configuredSources.isNotEmpty()
        if (hasDirectSource == hasNamedSources) {
            throw GradleException(
                "Configure exactly one of `codePath` or `namedSource(...)` for Embed Code.",
            )
        }

        val arguments = mutableListOf<String>()
        arguments.add("-mode=$executionMode")
        if (hasNamedSources) {
            arguments.add("-config-path=${writeNamedSourceConfiguration(configuredSources)}")
        } else {
            val sourceDirectory = codePath.get().asFile
            val documentationDirectory = docsPath.get().asFile
            logger.info(
                "Using source root `{}` and documentation root `{}`.",
                sourceDirectory,
                documentationDirectory,
            )
            arguments.add("-code-path=${sourceDirectory.absolutePath}")
            arguments.add("-docs-path=${documentationDirectory.absolutePath}")
            if (docIncludes.get().isNotEmpty()) {
                arguments.add("-doc-includes=${docIncludes.get().joinToString(",")}")
            }
            if (docExcludes.get().isNotEmpty()) {
                arguments.add("-doc-excludes=${docExcludes.get().joinToString(",")}")
            }
            arguments.add("-separator=${separator.get()}")
            arguments.add("-info=${info.get()}")
            arguments.add("-stacktrace=${stacktrace.get()}")
        }

        val executable = executableFile.get().asFile
        logger.info(
            "Starting Embed Code `{}` mode with executable `{}`.",
            executionMode,
            executable,
        )
        execOperations.exec { spec ->
            spec.executable(executable)
            spec.args(arguments)
            spec.setWorkingDir(processDirectory)
        }
        logger.info("Embed Code `{}` mode completed successfully.", executionMode)
    }

    /**
     * Writes the generated configuration used when named source roots are configured.
     */
    private fun writeNamedSourceConfiguration(configuredSources: Map<String, String>): Path {
        val normalizedSources = TreeMap<String, String>()
        for (source in configuredSources.entries) {
            var path = Paths.get(source.value)
            if (!path.isAbsolute) {
                path = workingDirectory.get().asFile.toPath().resolve(path)
            }
            path = path.normalize().toAbsolutePath()
            if (!Files.isDirectory(path)) {
                throw GradleException(
                    "Embed Code source `${source.key}` is not a directory: $path",
                )
            }
            normalizedSources[source.key] = path.toString()
        }

        val json = createConfigurationJson(
            normalizedSources,
            docsPath.get().asFile.absolutePath,
            docIncludes.get(),
            docExcludes.get(),
            separator.get(),
            info.get(),
            stacktrace.get(),
        )
        val configuration = temporaryDir.toPath().resolve("embed-code.json")
        try {
            Files.write(configuration, json.toByteArray(StandardCharsets.UTF_8))
        } catch (exception: IOException) {
            throw GradleException(
                "Could not write the generated Embed Code configuration to $configuration.",
                exception,
            )
        }
        logger.info(
            "Generated Embed Code configuration at `{}` for {} named source roots.",
            configuration,
            configuredSources.size,
        )
        return configuration
    }

}
